### PR TITLE
[CORE] [Clang]Cleanup clang-analyzer warnings

### PR DESCRIPTION
**Pull request by smuzaffar**
This PR fixes clang-analyzer warnings 
- https://cmssdt.cern.ch/SDT/jenkins-artifacts/ib-static-analysis/CMSSW_14_2_X_2024-10-03-1100/el8_amd64_gcc12/build-logs/
- https://cmssdt.cern.ch/SDT/jenkins-artifacts/ib-static-analysis/CMSSW_14_2_X_2024-10-03-1100/el8_amd64_gcc12/llvm-analysis/index.html

### CODE REVIEW

**Comment by makortel** [comment]
FYI @gartung for the static analyzer (changes look good to me)

File: Utilities/StorageFactory/src/LocalFileSystem.cc [hunk 1]
```diff
@@ -361,6 +361,7 @@ LocalFileSystem::FSInfo *LocalFileSystem::findMount(const char *path,
 
     prev_paths.push_back(path);
     LocalFileSystem::FSInfo *new_best = findMount(fullpath, &sfs2, &s2, prev_paths);
+    free(fullpath);
```
Comment (makortel) [inline Utilities/StorageFactory/src/LocalFileSystem.cc:364]: How about using a `unique_ptr` for the `fullpath`, e.g.
```cpp
std::unique_ptr<char, decltype(std::free)> fullpath{realpath(best->origin, nullptr), std::free};

if (!fullpath)
  fullpath.reset(strdup(best->origin));
```
removing the `free(fullpath)` calls and replacing the accesses with `fullpath.get()`?

In principle the code could be updated to use `std::filesystem`, but that sounds a lot more work that would have to be done with care.

Comment (smuzaffar) [reply Utilities/StorageFactory/src/LocalFileSystem.cc:364]: @makortel , I have tried `std::unique_ptr` and after replacing `fullpath` to`fullpath.get()` and few other places (`lstat, statfs, findMount` need `const char*`) I got build error
```
In file included from /cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/bits/unique_ptr.h:36,
                 from /cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/memory:75,
                 from /cvmfs/cms-ib.cern.ch/sw/x86_64/week0/el8_amd64_gcc12/cms/cmssw/CMSSW_14_2_X_2024-10-06-2300/src/FWCore/MessageLogger/interface/MessageLogger.h:23,
                 from src/Utilities/StorageFactory/src/LocalFileSystem.cc:4:
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/tuple: In instantiation of 'struct std::_Head_base<1, void(void*) noexcept, false>':
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/tuple:416:12:   required from 'struct std::_Tuple_impl<1, void(void*) noexcept>'
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/tuple:258:12:   required from 'struct std::_Tuple_impl<0, char*, void(void*) noexcept>'
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/tuple:981:11:   required from 'class std::tuple<char*, void(void*) noexcept>'
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/bits/unique_ptr.h:224:27:   required from 'class std::__uniq_ptr_impl<char, void(void*) noexcept>'
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/bits/unique_ptr.h:255:12:   required from 'struct std::__uniq_ptr_data<char, void(void*) noexcept, false, false>'
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/bits/unique_ptr.h:275:33:   required from 'class std::unique_ptr<char, void(void*) noexcept>'
src/Utilities/StorageFactory/src/LocalFileSystem.cc:341:48:   required from here
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02858/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/include/c++/12.3.1/tuple:238:13: error: data member 'std::_Head_base<1, void(void*) noexcept, false>::_M_head_impl' invalidly declared function type
  238 |       _Head _M_head_impl;
      |             ^~~~~~~~~~~~
src/Utilities/StorageFactory/src/LocalFileSystem.cc: In member function 'edm::storage::LocalFileSystem::FSInfo* edm::storage::LocalFileSystem::findMount(const char*, statfs*, stat*, std::vector<std::__cxx11::basic_string<char> >&) const':
src/Utilities/StorageFactory/src/LocalFileSystem.cc:341:100: error: no matching function for call to 'std::unique_ptr<char, void(void*) noexcept>::unique_ptr(<brace-enclosed initializer list>)'
  341 |     std::unique_ptr<char, decltype(std::free)> full_path{realpath(best->origin, nullptr), std::free};

```

If I change 
```
std::unique_ptr<char, decltype(std::free)> fullpath{realpath(best->origin, nullptr), std::free};
```
to
```
std::unique_ptr<char, decltype(std::free) *> fullpath{realpath(best->origin, nullptr), std::free};
```

then build error go away.  So is  `std::unique_ptr<char, decltype(std::free) *>` the correct thing touse in this case?

**Comment by smuzaffar** [comment]
please test

File: Utilities/StorageFactory/src/LocalFileSystem.cc (hunk 1 above)
Comment (makortel) [reply Utilities/StorageFactory/src/LocalFileSystem.cc]: > So is `std::unique_ptr<char, decltype(std::free) *>` the correct thing touse in this case?

Ah, yes, it is the correct type. Sorry for omitting the `*`.

File: Utilities/StorageFactory/src/LocalFileSystem.cc [hunk 2]
```diff
@@ -168,7 +168,7 @@ LocalFileSystem::FSInfo *LocalFileSystem::initFSInfo(void *arg) {
   i->fsname = static_cast<char *>(memcpy(p += infolen, m->mnt_fsname, fslen));
   i->type = static_cast<char *>(memcpy(p += fslen, m->mnt_type, typelen));
   i->dir = static_cast<char *>(memcpy(p += typelen, m->mnt_dir, dirlen));
-  i->origin = static_cast<char *>(memcpy(p += dirlen, m->mnt_fsname, originlen));
+  [[clang::suppress]] i->origin = static_cast<char *>(memcpy(p += dirlen, m->mnt_fsname, originlen));
```
[the remark this answers is not in the document]
Comment (makortel) [reply Utilities/StorageFactory/src/LocalFileSystem.cc:171]: @smuzaffar Do you prefer to keep the `[[clang::suppress}}` here?

[the remark this answers is not in the document]
Comment (smuzaffar) [reply Utilities/StorageFactory/src/LocalFileSystem.cc:171]: No really @makortel .

**Comment by smuzaffar** [comment]
please test

**Comment by makortel** [comment]
Comparison differences are related to https://github.com/cms-sw/cmssw/issues/39803

**Comment by makortel** [comment]
+core

**Comment by mandrenguyen** [comment]
+1

**Pull request merged**